DOCS - ACL migration (#5418)
treeverse/lakeFS: docs/reference/ACL-migration.md

### The new ACL model
As mentioned in our [Security Update](https://docs.lakefs.io/posts/security_update.html#whats-changing) the coming lakeFS version introduces a simplified authorization mechanism based on ACLs.

### Permissions

> In the ACL model Every group will have a permission attached to it,
> lakeFS provides the following permissions:

Reviewer: ```suggestion
The new ACL model defines permissions at the group level, we'll review the new permissions compared to their previous representation:
```

> | Permission | Allows | Existing Group |

Reviewer: Maybe we should name the "Existing Group" differently (e.g Previous Group Name) and add a comment? 
I find it a bit confusing

Reviewer: Also, I think we should compare it to the permission themselves and not the groups; we can't really compare it to groups as they might have changed, while the permissions themselves weren't.

| **Super** | Allows all operations except auth. | SuperUsers (with changes) |
| **Admin** | Allows all operations. | Admins |

### Scopes

When granted to a group, permissions Read, Write, and Super may be scoped to a set of repositories.
Admin includes global abilities that apply across repos and cannot be scoped to a set of repositories

### ACL

You can now attach Permissions and scope them to groups in the Groups page
There are 4 default groups, named after the 4 permissions. Each group is global (applies for all repositories).
The group list initially would be

| Group ID    | Default Permission | Created at       | Repositories |
|-------------|--------------------|------------------|--------------|
| **Read**    | Read               | \<Creation date> | All          |
| **Write**   | Write              | \<Creation date> | All          |
| **Super**   | Super              | \<Creation date> | All          |
| **Admin**   | Admin              | \<Creation date> | All          |

In the UI:

![Groups page has a dropdown to edit permission for each
group]( ../assets/img/ACL-groups-with-perms.png)

### The migration

> Upgrading the lakeFS version will require running a migrating to our ACL authorization model in case:

Reviewer: ```suggestion
Upgrading the lakeFS version will require migrating to the new ACL authorization model in case:
```

Reviewer: Isn't it required anyway? and requires manual intervention if the following actions took place?

Reviewer 2: Yup, always required.  (Technically it will might with no update, but you will be unable to edit any policies or even see what they are.)

> - One or more of the basic poicies was modified

Reviewer: ```suggestion
- One or more of the basic policies was modified
```

> - New policies where defined

Reviewer: ```suggestion
- New policies were created
```

> - Users have attached policies

Reviewer: ```suggestion
- Users have attached policies directly
```

In order to run the migration run:
```

> TODO - Add the command

Reviewer 2: ```sh
lakefs migrate auth-acl
```

```

> The command defaults to dry-run,
> We recommend running the script only after running the dry-run and going over the warnings
> In order to apply the migration you will need  `--yes`  flag

Reviewer 2: ```suggestion
The command defaults to dry-run.
Running the script and go over the warnings about widened or otherwise changed permissions.
To apply the migration, re-run with the  `--yes`  flag
```

Author: @arielshaqed , I don't understnad what you meant in:
> Running the script and go over the warnings about widened or otherwise changed permissions.

The upgrade will ensure that the 4 default groups exist, and modify existing groups to fit into the new ACLs model:
-  When creating the 4 default global groups: if another group exists and has the desired name, upgrading will rename it by appending ".orig". So after upgrading the 4 default global groups exist, with these known names.
- For any group, upgrading configured policies follows these rules, possibly increasing access:
    1. Any "Deny" rules are stripped, and a warning printed.
    2. "Manage own credentials" is added.

>     3. If any actions outside of "fs:" are allowed, the group becomes an Admin group, a warning is printed, and no further changes apply.

Reviewer 2: ```suggestion
    3. If any actions outside of "fs:" and manage own credentials are allowed, the group becomes an Admin group, a warning is printed, and no further changes apply.
```

>     4. The upgrade script unifies repositories: If a permission applies to a set of repositories with a wildcard, permissions are unified to all repositories. Otherwise they apply to the list of all repositories, in all the policies.

Reviewer 2: ```suggestion
    4. The upgrade script unifies repositories: If a resource applies to a set of repositories with a wildcard, permissions are unified to all repositories. Otherwise they apply to the list of all repositories, in all the policies.
    5. Any paths appearing in any resource are removed.
```

>     5. The upgrade script unifies actions: it selects the least permission of Read, Write, Super that contains all of the allowed actions. If any allowed action contains a wildcard this still makes sense.

Reviewer: > If any allowed action contains a wildcard this still makes sense.

I'm a bit confused with this sentence and how it relates to the full sentence, care to explain?

Reviewer 2: Just remove the second sentence: it is poorly phrased.  Consider adding instead "Any wildcards appearing in actions are expanded and handled."

>        All warnings link to support@ and to Slack.

Reviewer: Redundant @? also, what does this mean?

If this means that we'll support the community with the upgrade, I suggest writing this clearly. for instance:
For any question or concern during the upgrade, don't hesitate to get in touch with us through Slack or Support email.

Reviewer 2: "support@treeverse.io" :-)
